### PR TITLE
fix[l2geth]: have successful deposits return a status of 1

### LINKED ISSUE #497: ETH deposits all return status 0 (Revert)
Despite succeeding (ETH actually deposits and the correct events are emitted), ETH deposits show status: 0 when they are processed on L2.

**Steps to reproduce the behavior:**
1. Run native eth integration tests
2. Check getTransactionReceipt on the L2 tx hash of a completed ETH deposit.

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Adds a temporary fix to `l2geth` to ensure that successful L1 => L2 messages will return a status of 1 ("success") instead of 0 ("failure"). Ideally we shouldn't be treating L1 => L2 messages in a special way and instead use the technique attempted in https://github.com/ethereum-optimism/optimism/pull/897, but that technique requires we make an update to the `OVM_L2CrossDomainMessenger` and we need a short-term fix.

**Metadata**
- Fixes #497

### CODE REVIEW

File: l2geth/core/vm/evm.go [hunk 1]
```diff
@@ -323,6 +326,15 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 		}
 	}
 
+	// We need to be able to show a status of 1 ("no error") for successful L1 => L2 messages.
+	// Unfortunately, the current way we figure out the success of the transaction (by parsing the
```
Comment (tynes) [inline l2geth/core/vm/evm.go:330]: We should reword this comment to not use words like `Unfortunately` - we should just explain what the solution is. We should also remove the words `temporary hack`

**Review by ben-chain — CHANGES_REQUESTED** [review]
This LGTM as far as the core logic.   I would recommend that we add an integration test showing a failed deposit to make sure we have that case covered in geth too.  Also, once this is merged, we should document the current approach and have an issue to track using returndata in the next regenesis.

File: l2geth/core/vm/evm.go [hunk 2]
```diff
@@ -323,6 +326,15 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 		}
 	}
 
+	// We need to be able to show a status of 1 ("no error") for successful L1 => L2 messages.
+	// The current way we figure out the success of the transaction (by parsing the
+	// returned data) would require an upgrade to the contracts that we likely won't make for a
+	// while. As a result, we'll use this mechanism where we set IsOkL1ToL2Message = true if
+	// we detect an L1 => L2 message that didn't revert. Initially we just set the value to false.
+	if evm.depth == 0 {
```
Comment (tynes) [inline l2geth/core/vm/evm.go:334]: The zero value for a boolean is false so this is just being explicit

**Comment by tynes** [comment]
I agree that a test showing that it is possible to have a reverted deposit is a good idea. This calls for a standardization around returndata so that alternative bridges can "just work"

**Comment by smartcontracts** [comment]
>  I would recommend that we add an integration test showing a failed deposit

@ben-chain how should we define a "failed" deposit? Is it based on the success of the call that the messenger makes? Or based on errors inside of the call from the EM to the messenger itself?

**Comment by ben-chain** [comment]
> @ben-chain how should we define a "failed" deposit? Is it based on the success of the call that the messenger makes? Or based on errors inside of the call from the EM to the messenger itself?

I think the former is the way to go.  I like the pattern that we already have going with the entrypoint--reversions are captured via `return abi.encode(success, data)`, and any exceptions surfaced directly to the EM are system-level bugs.

File: l2geth/core/vm/evm.go (hunk 2 above)
Comment (smartcontracts) [reply l2geth/core/vm/evm.go:340]: I didn't know if geth ever reuses the same `evm` object to do multiple executions, I felt it was safer to reset this for each new call. I can delete if you think this is unnecessary.

File: l2geth/core/vm/evm.go [hunk 3]
```diff
@@ -369,8 +389,35 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
...
+	}
+
 	ret, err = run(evm, contract, input, false)
 
+	// If all of these very particular conditions hold then we're guaranteed to be in a successful
+	// L1 => L2 message. It's not pretty, but it works. Broke this out into a series of checks to
+	// make it a bit more legible.
+	if evm.Context.IsL1ToL2Message && evm.depth == 3 {
+		var isValidMessageTarget = true
+		// 0x420... addresses are not valid targets except for the ETH predeploy.
```
Comment (ben-chain) [inline l2geth/core/vm/evm.go:406]: 👌

File: packages/core-utils/src/watcher.ts [hunk 4]
```diff
@@ -98,19 +105,23 @@ export class Watcher {
 
     // Message has yet to be relayed, poll until it is found
     return new Promise(async (resolve, reject) => {
-      layer.provider.on(filter, async (log: any) => {
+      const handleEvent = async (log: any) => {
```
Comment (ben-chain) [inline packages/core-utils/src/watcher.ts:108]: Assuming this was introduced by adding the failed message event type, and we just now caught this from now having tested failed messages.  If so, should a watcher patch be included in the changelog?

**Review by ben-chain — APPROVED** [review]
One question about the changelog but LGTM!

**Comment by tynes** [comment]
This is based on develop but requires a resync if we want the receipts to be populated correctly. If different infrastructure providers updated at different heights, then it would confusing for users - getting a receipt from the sequencer could show a different status than what Etherscan shows.

We should probably wait until the next regenesis for this one. Thoughts @K-Ho?

**Comment by smartcontracts** [comment]
> This is based on develop but requires a resync if we want the receipts to be populated correctly. If different infrastructure providers updated at different heights, then it would confusing for users - getting a receipt from the sequencer could show a different status than what Etherscan shows.
> 
> We should probably wait until the next regenesis for this one. Thoughts @K-Ho?

This PR was specifically designed so it could be merged without a regenesis. I really think it's worth trying to get Etherscan to resync using develop once this is merged. But if for whatever reason we want to wait for a regenesis we should do this PR the "correct" way and update the data returned by the L2CrossDomainMessenger so we don't have to do this low-level hack.

File: l2geth/core/vm/evm.go (hunk 2 above)
Comment (tynes) [reply l2geth/core/vm/evm.go:341]: I agree that having it better, its good to be explicit. Would be surprised if an evm is used more than once

**Comment by K-Ho** [comment]
This will be a one-time re-sync and should be totally fine. Basically we merge this PR, then Etherscan and Quiknode will have to re-sync their replicas. During this day, they may be out of sync, but not many people are using Quiknode yet, so not a big deal. After the re-sync, both of their replicas will be displaying the correct statuses for all deposits and we can proceed. Yay all good!

Other note is that the sequencer's historical deposits will have status 0, but we're blacklisting access to even querying `eth_getTransactionReceipt` on the sequencer, so that shouldn't be a problem either

File: integration-tests/test/basic-l1-l2-communication.spec.ts [hunk 5]
```diff
@@ -28,48 +31,101 @@ describe('Basic L1<>L2 Communication', async () => {
...
+        env.l2Wallet.address
+      )
+      expect(await L1SimpleStorage.value()).to.equal(value)
+      expect((await L1SimpleStorage.totalCount()).toNumber()).to.equal(1)
+    })
   })
 
-  it('should deposit from L1 -> L2', async () => {
-    const value = `0x${'42'.repeat(32)}`
+  describe('L1 => L2', () => {
```
Comment (tynes) [inline integration-tests/test/basic-l1-l2-communication.spec.ts:74]: Adding tests for unsuccessful messages to dead addresses and 0x42 addrs would be nice to add here

**Review by tynes — APPROVED** [review]

**Pull request merged**